kirkwood: add support for D-Link DNS-325 A1

SimplyKot opened the pull request:
The D-Link DNS-325 A1 is NAS with 2x HDD (3.5 inch) bays, 1x 10/100/1000Base-T port and 1x USB 2.0 port

Device specification:

    SoC: Marvell 88F6281-A1 1.2 GHz
    RAM: 256 MB (2xSEC K4T1G084QF-HCF7:128Mx8 @400 CL6)
    Flash: 128 MB (SAMSUNG 946 K9F1G08U0B PCB0)
    SATA: 2x internal SATA II drives
    Ethernet: 1x Gigabit (Marvell 88E1116R-NNC1)
    LED: 7x (white: Power, white/red: Right SATA Activity, USB Activity, Left SATA Activity)
    Key: 3x (Power, Reset, USB Copy/Unmount)
    Serial: 5 pin header (RXD,GAP,3.3V,GND,TXD), (115200,8,N,1), 3.3V TTL
    USB ports: 1x USB 2.0

Flash instruction:

NOTE: this process uses a serial connection. It will upgrade the bootloader and reset the bootloader environment variables

USB flash stick setup

    Format to FAT32 without mbr
    Copy these files to USB flash stick: dlink_dns-325-a1-initramfs-uImage dlink_dns-325-a1-squashfs-factory.bin dlink_dns-325-a1-squashfs-sysupgrade.bin u-boot.kwb (from u-boot-dns325 directory)

NAS setup

    Connect LAN cable between router and NAS device
    Connetc USB flash stick to NAS device
    Connect serial to NAS device

Boot from new u-boot and install it

    $ kwboot -p -b u-boot.kwb -B115200 -t /dev/ttyUSB0
    Power on NAS
    After download copleted and booted NAS device with new u-boot stop in bootloader by pressing any key
    => usb start
    => fatload usb 0 0x1000000 /u-boot.kwb
    => nand erase.part u-boot
    => nand write 0x1000000 u-boot ${filesize}
    => reset

Update MAC address in u-boot env

    Stop in bootloader by pressing any key
    Get your MAC address from label on chassis
    => setenv ethaddr XX:XX:XX:XX:XX:XX
    => saveenv

Install OpenWrt by flashing factory image from u-boot

    => usb start
    => fatload usb 0 0x1000000 /dlink_dns-325-a1-squashfs-factory.bin
    => nand erase.part ubi
    => nand write 0x1000000 ubi ${filesize}
    => reset

Based on https://github.com/openwrt/openwrt/pull/9296

Signed-off-by: Konstantin Meshkov <skyline@mail.ru>

SimplyKot commented:
@wigyori Can I ask you to check my code if you have time?

Neustradamus commented:
@SimplyKot: Good job!

Maybe you can check other D-Link DNS-320/320L/321/323/327L and DNR-322L too?

It replaces:
- https://github.com/openwrt/openwrt/pull/17918

Neustradamus commented:
@SimplyKot: Good job! :)

Maybe you can check other D-Link DNS-320/320L/321/323/327L and DNR-322L too?

A closed PR about DNS-320L:
- https://github.com/openwrt/openwrt/pull/15433

DNS-313 and DNS-315 too?

Here a @linusw website page about D-Link DNS-313:
- https://www.df.lth.se/~triad/krad/dlink-dns-313/

It replaces:
- https://github.com/openwrt/openwrt/pull/17918
- https://github.com/openwrt/openwrt/pull/17912
- https://github.com/openwrt/openwrt/pull/9296

Neustradamus commented:
Dear all, @lentinj, @davidedg, @Elenedeath, @bjaglin, @awanga, @NicolasBernaerts, @lecocqe, @gpancot, @barnumbirr, @katharsis1312, @alexey85, @jctim, @Cloudxtreme, @qsypoq, @alxbun, @bishop41466, @coolshou, @mpalmer, @juergh, @djon2003, @hongkiaong, @tgragnato, @twyatt, @pdxrwilliams, @martignlo, @scus1, @ggirou, @pticon, @stefanocudini, @ggrisandi, @avoidik, @rmtsrc, @dominis, @wesavetheworld, @martin-vi, @dg14, @APCCV, @jeremiehuchet, @enkiusz, @cbenning, @wigyori, @andyboeh, @5p0ng3b0b, @JonasKress, @utix, @TomasBouda, @jszigetvari, @kekstlt, @cheshaneko, @cheshaneko-old,

Can you look this PR and previous comments?

Thanks in advance.

utix commented on package/boot/uboot-kirkwood/Makefile:
```diff
@@ -112,6 +117,7 @@ endef
 
 UBOOT_TARGETS := \
 	dns320l \
+  dns325 \
```
Indent issue

utix commented on package/boot/uboot-kirkwood/patches/011-arm-kirkwood-dns325-Correct-CONFIG_NR_DRAM_BANKS-parameter.patch:
```diff
@@ -0,0 +1,32 @@
+ CONFIG_IDENT_STRING="\nD-Link DNS-325"
+ # CONFIG_SYS_MALLOC_F is not set
+ CONFIG_BOOTDELAY=3
+-- 
```
trailing spaces

utix reviewed (approved):
Pedantic comment.
Look good to me

jszigetvari commented:
> DNS-313 and DNS-315 too?
> 
> Here a @linusw website page about D-Link DNS-313:
> 
>     * https://www.df.lth.se/~triad/krad/dlink-dns-313/
> 
> 
> OpenWrt doc about DNS-313:
> 
>     * https://openwrt.org/toh/d-link/dns-313

@Neustradamus Sorry, I can't help you with that anymore. I sold off my device like 10 years ago.

NicolasBernaerts commented:
If it helps, I've published my debian buster installation manual for DNS-325.
But I don't have that fantastic device any more.

https://github.com/NicolasBernaerts/dns-32x/tree/master/debian/dns325/buster

APCCV commented:
This looks familiar. We are all working from the same online materials.

I might have reinvented the wheel for the DNS-320-A1. Ended up with something similar but not quite identical to this PR. Haven't created the PR yet, but posted the repo on forums [here](https://forum.openwrt.org/t/openwrt-for-d-link-dns320-a1/224382). Some reference links there for 320 that may apply to 325 too. Need to clean what I have (especially u-boot env) and retest before I PR it.

Somehow I ended up with a different set of patches for u-boot. The devices are very similar but not identical. I'm sure I tried something like what is in this commit for the 320 and it didn't boot or ethernet didn't work. I handled changes to dts diferently, overlaying on top of the upstream version instead of patching it. Can't remember why, probably tried patching and I didn't get it to work - had to do with the includes. End result might be the same.

The 320 relies on a shell+cron solution for fan control that really bothers me. Cron spams the log like there's no tomorrow and the shell script has a race condition when querying temperature on the 320. Maybe this is not an issue with the 325. I wrote a small C program that works better IMO. If the 325 uses ttyS1 to get temperature maybe we should use what I have for both. Small adjustments may be necessary (temp is reported in two different spots on serial response, I only implemented the one I need).

What I would really like to figure out is how to flash without serial. Need to test it (waiting for a special USB-TTL adapter before I can try it out) but I'm hoping that, starting from stock, then Alt-F (which might allow to change the u-boot env), then to booting from USB and from there flashing u-boot and ubi. Riskier but a lot friendlier for more OpenWrt tinkerers.

lentinj commented on package/boot/uboot-kirkwood/patches/100-dns325.patch:
```diff
@@ -0,0 +1,75 @@
+ 	kw_gpio_set_blink(DNS325_GPIO_LED_POWER , 1);
+ 
+ 	kw_gpio_set_value(DNS325_GPIO_SATA0_EN , 1);
++	kw_gpio_set_value(DNS325_GPIO_SATA1_EN , 1);
```
The reason SATA1 wasn't enabled at this point was to stagger the drive spin up and avoid overloading the PSU, if I remember correctly. U-boot spins up the first drive early, and Linux spins up the next.

I'm sure there's a better way of doing this, but enabling both drives in u-boot wouldn't allow for this at all I think?

SimplyKot replied:
Thank you. 
A am already recompiled u-boot as you say and now right HDD starts after left after 5 seconds.
```
[    4.165467] ata1.00: configured for UDMA/133
[    4.170628] scsi 0:0:0:0: Direct-Access     ATA      WDC WD5000AAKX-0 1H15 PQ: 0 ANSI: 5
[    4.181911] sd 0:0:0:0: [sda] 976773168 512-byte logical blocks: (500 GB/466 GiB)
[    4.189592] sd 0:0:0:0: [sda] Write Protect is off
[    4.194597] sd 0:0:0:0: [sda] Write cache: enabled, read cache: enabled, doesn't support DPO or FUA
[    4.203933] sd 0:0:0:0: [sda] Preferred minimum I/O size 512 bytes
[    4.220097] sd 0:0:0:0: [sda] Attached SCSI disk
[    9.748900] ata2: link is slow to respond, please be patient (ready=0)
[    9.888916] ata2: SATA link up 3.0 Gbps (SStatus 123 SControl F300)
[    9.919192] ata2.00: ATA-8: SAMSUNG HD322GJ, 1AR10001, max UDMA/133
[    9.925533] ata2.00: 625142448 sectors, multi 0: LBA48 NCQ (depth 31/32)
[    9.959201] ata2.00: configured for UDMA/133
[    9.964490] scsi 1:0:0:0: Direct-Access     ATA      SAMSUNG HD322GJ  0001 PQ: 0 ANSI: 5
[    9.975918] sd 1:0:0:0: [sdb] 625142448 512-byte logical blocks: (320 GB/298 GiB)
[    9.983644] sd 1:0:0:0: [sdb] Write Protect is off
[    9.988653] sd 1:0:0:0: [sdb] Write cache: enabled, read cache: enabled, doesn't support DPO or FUA
[    9.997991] sd 1:0:0:0: [sdb] Preferred minimum I/O size 512 bytes
[   10.019140] sd 1:0:0:0: [sdb] Attached SCSI disk 
```

BulldozerBSG replied:
Hi @SimplyKot, @lentinj,
Note: In the upstream release of u-boot both disks are powered on startup
Enabled by this commit https://source.denx.de/u-boot/u-boot/-/commit/fec546264797022be2d70c441bd2d07fe9045056

I think... You will always have situations on a running system where two disks first go to sleep and then wake up at the same time. It seems to me that delayed startup does not help much with overloading PSU.

But perhaps the smaller patch the better for support )

lentinj commented:
> The 320 relies on a shell+cron solution for fan control that really bothers me. Cron spams the log like there's no tomorrow and the shell script has a race condition when querying temperature on the 320. Maybe this is not an issue with the 325.

The 325 has a much more normal lm75 temperature sensor hanging off I2C. The serial-connected temperature reading is DNS-320 only problem (possibly DNS-320L too?).

lentinj commented:
> What I would really like to figure out is how to flash without serial. Need to test it (waiting for a special USB-TTL adapter before I can try it out) but I'm hoping that, starting from stock, then Alt-F (which might allow to change the u-boot env), then to booting from USB and from there flashing u-boot and ubi. Riskier but a lot friendlier for more OpenWrt tinkerers.

There's certainly nothing stopping you changing the u-boot-env from Alt-F, or using the same hook that Alt-F uses, with some statically compiled binaries. I considered trying something along these lines back in the day, but didn't want to be responsible for any associated bricking :)

andyboeh commented:
The DNS320L has an MCU sitting at /dev/ttyS1 for fan control, temperature reading, device shut down, RTC, device power on ...

The respective protocol has been reversed about 10 years ago and with 24.10.0, official support for the DNS-320L including the MCU daemon has landed.

I don't know (remember) what the differences between DNS-320 and DNS-320L are?

APCCV commented:
> DNS-320 only problem

My luck.  :)

Hey, it's you! Can't thank you enough for all the information you posted online about these devices. Your [website](https://jamie.lentin.co.uk/devices/dlink-dns325/) was super helpful to get the 320 going.

APCCV commented:
> differences between DNS-320 and DNS-320L

This was of help for me.
[320](http://dns323.kood.org/dns-320)
[325](http://dns323.kood.org/dns-325)
[302L](http://dns323.kood.org/dns-320l)
Also [this](https://jamie.lentin.co.uk/devices/dlink-dns325/) - 320L mentioned below 320/325 comparison.

The 320 is the slowest and simplest of the three. Probably the first version of the family. 325 has the same SOC but twice more RAM and a different PMU (resulting in a better wired temp sensor as mentioned above). 
[Here](https://archlinuxarm.org/forum/viewtopic.php?t=9627&start=0) is a fun page on overclocking an 800MHz Pogo Plug that uses the same SOC which maybe would make a 320 run at 1GHz.
320L is newer than the other two and has a different SOC plus the same PMU that the 320 has, it seems, but properly wired.
Gigabit chipset may also differ.

I guess in the end it means a bit of a nightmare in u-boot build and settings for each. I mixed and matched many different 320/325/320L u-boot configs and patches until I found something that worked for the 320. I don't think it's possible to have a single config in OpenWrt to support them all.

Neustradamus commented:
Dear all,

I have forgotten to add previously:
@Tho85, @szemtiv, @iga566, @Le-candide, @pcs3rd, @Darkvater, @abdelkader, @blackhorus, @gpancot, @binlab, @charliemartinez

Can you look this PR and previous comments?

In more, we can look people who have worked on OpenWrt uboot-kirkwood:
@lperkov, @p-wassi, @blogic, @nbd168, @bobafetthotmail, @mkresin, @hauke, @ynezz, @CHKDSK88, @kapieyow, @berenyibalazs, @kuanyili, @Ansuel, @1715173329, @neheb

Thanks in advance.

wigyori commented:
@SimplyKot - good to see other folks working on giving these boxes a new life. :)

A few comments/nitpicks:

 - In [108-dns325.patch](https://github.com/openwrt/openwrt/pull/17924/commits/4505b1a824dc8b8e98ec719ac14585c893fe3884#diff-91a21ca18da44e6220f6dfa24b1ecf42de0f3ed7bf4b42ee32a9300d53897f37), you are updating kirkwood-dnskw.dtsi, which is included from kirkwood-dns320.dts as well. Although we're not supporting DNS-320(A1) at the moment IIRC, your patch implies that the updated partition table will be used on the DNS-320 - which is likely, but not necessarily will be the case. I'd rather delete-node the partitions in kirkwood-dns325.dts and re-add them with the new sizes.
 - Also, I'd bring out the hunks updating kirkwood-dnskw.dtsi into a separate patch file for clarity.
 - [image/Makefile](https://github.com/openwrt/openwrt/pull/17924/commits/4505b1a824dc8b8e98ec719ac14585c893fe3884#diff-a8080bc7f3817a35be884486b08edaf735f267e21d5051b140de1c792a56ab82) - I would not add the "block-mount mdadm partx-utils fdisk kmod-md-linear" packages - neither of the other NASes have these. To be fair, while working on the Stora and the DNS-320L, I also had a similar idea to make the users' life easier, but then dropped it to match up with the other NASes.
 - Although there is ample information about this NAS on the interwebz, please make sure to create a wiki page about it on openwrt.org, with some links to the sites listed above.
 - Other than that, a couple indent/whitespace/typos, but @utix have gone through these already.

Thanks!

BulldozerBSG commented:
Hi @SimplyKot,

 I tried build the firmware (24.10.0 + PR, maybe this is not relevant for the master branch) with this PR and got errors:
 ```
 Collected errors:
 * pkg_hash_check_unresolved: cannot find dependency kmod-thermal for kmod-hwmon-gpiofan
 * pkg_hash_fetch_best_installation_candidate: Packages for kmod-hwmon-gpiofan found, but incompatible with the architectures configured
 * opkg_install_cmd: Cannot install package kmod-hwmon-gpiofan.
 * pkg_hash_check_unresolved: cannot find dependency kmod-thermal for kmod-hwmon-lm75
 * pkg_hash_fetch_best_installation_candidate: Packages for kmod-hwmon-lm75 found, but incompatible with the architectures configured
 * opkg_install_cmd: Cannot install package kmod-hwmon-lm75.
 ```
 I'm not sure but found this commit https://github.com/openwrt/openwrt/commit/ccbdb212f36f6db25062c5b50c9729842e60fa24
 Since kernel version 6.1, it is required to add a package `kmod-thermal` as a dependency, for example:
 ```
 diff --git a/target/linux/kirkwood/image/Makefile b/target/linux/kirkwood/image/Makefile
index 1d90930e43..e5856cc516 100644
--- a/target/linux/kirkwood/image/Makefile
+++ b/target/linux/kirkwood/image/Makefile
@@ -193,7 +193,7 @@ define Device/dlink_dns-325-a1
   DEVICE_MODEL := DNS-325
   DEVICE_VARIANT := A1
   DEVICE_DTS := kirkwood-dns325
-  DEVICE_PACKAGES := kmod-ata-marvell-sata kmod-hwmon-lm75 kmod-hwmon-gpiofan \
+  DEVICE_PACKAGES := kmod-ata-marvell-sata kmod-hwmon-lm75 kmod-hwmon-gpiofan kmod-thermal \
        kmod-usb-ledtrig-usbport kmod-gpio-button-hotplug block-mount mdadm \
   kmod-md-linear kmod-usb-storage kmod-fs-ext4 e2fsprogs partx-utils fdisk \
   
 ```

glassum commented:
I'm quite eager to experiment with my DNS-323, will some version for 325 at least start on it? :)

SimplyKot commented:
> kmod-thermal \

Just made the corrections you suggested. Thank you.

CHKDSK88 commented:
IMO it should be rebased to current main and tested with 6.12.

SimplyKot commented:
> IMO it should be rebased to current main and tested with 6.12.

Thanks for the advice.

I'll start working on this update this week.

SimplyKot commented:
> IMO it should be rebased to current main and tested with 6.12.

Rebased to main and now testing.
<img width="900" height="491" alt="изображение" src="https://github.com/user-attachments/assets/ffe07efd-ed71-4c9d-aaee-bcdcda8939f7" />

SimplyKot commented:
> mhhh refresh uboot patches and split 108 to one for dtsi and one for the dts (or even provide it outside a patch as a dedicated file)
> 
> Also for kernel patch at least lets make proper git patch with commit title and description

Thanks for the advice.

I've split patch 108 into separate patches for dts and dtsi, and added a commit description.

Neustradamus commented:
@SimplyKot: You have closed the PR :/

Ansuel commented:
@SimplyKot thanks a lot, maybe 2 last thing:
- Use different number for patch 108 (call the other 109)
- Is it possible to make formal git patch for patch 108 (and 109) ? (with sob, commit title and commit description?)

SimplyKot committed:
kirkwood: add support for D-Link DNS-325 A1

The D-Link DNS-325 A1 is NAS with 2x HDD (3.5 inch) bays, 1x 10/100/1000Base-T port and 1x USB 2.0 port

Device specification:

    SoC: Marvell 88F6281-A1 1.2 GHz
    RAM: 256 MB (2xSEC K4T1G084QF-HCF7:128Mx8 @400 CL6)
    Flash: 128 MB (SAMSUNG 946 K9F1G08U0B PCB0)
    SATA: 2x internal SATA II drives
    Ethernet: 1x Gigabit (Marvell 88E1116R-NNC1)
    LED: 7x (white: Power, white/red: Right SATA Activity, USB Activity, Left SATA Activity)
    Key: 3x (Power, Reset, USB Copy/Unmount)
    Serial: 5 pin header (RXD,GAP,3.3V,GND,TXD), (115200,8,N,1), 3.3V TTL
    USB ports: 1x USB 2.0

Flash instruction:

NOTE: this process uses a serial connection. It will upgrade the bootloader and reset the bootloader environment variables

USB flash stick setup

    Format to FAT32 without mbr
    Copy these files to USB flash stick: dlink_dns-325-a1-initramfs-uImage dlink_dns-325-a1-squashfs-factory.bin dlink_dns-325-a1-squashfs-sysupgrade.bin u-boot.kwb (from u-boot-dns325 directory)

NAS setup

    Connect LAN cable between router and NAS device
    Connetc USB flash stick to NAS device
    Connect serial to NAS device

Boot from new u-boot and install it

    $ kwboot -p -b u-boot.kwb -B115200 -t /dev/ttyUSB0
    Power on NAS
    After download copleted and booted NAS device with new u-boot stop in bootloader by pressing any key
    => usb start
    => fatload usb 0 0x1000000 /u-boot.kwb
    => nand erase.part u-boot
    => nand write 0x1000000 u-boot ${filesize}
    => reset

Update MAC address in u-boot env

    Stop in bootloader by pressing any key
    Get your MAC address from label on chassis
    => setenv ethaddr XX:XX:XX:XX:XX:XX
    => saveenv

Install OpenWrt by flashing factory image from u-boot

    => usb start
    => fatload usb 0 0x1000000 /dlink_dns-325-a1-squashfs-factory.bin
    => nand erase.part ubi
    => nand write 0x1000000 ubi ${filesize}
    => reset

Based-on: #9296
Signed-off-by: Konstantin Meshkov <skyline@mail.ru>
[ refresh uboot patches ]
Link: https://github.com/openwrt/openwrt/pull/17924
Signed-off-by: Christian Marangi <ansuelsmth@gmail.com>

SimplyKot committed:
kirkwood: add aliases and new partition for support D-Link DNS-325 A1

- Added fan0 alias
- Implemented LED aliases and thermal zones
- Restructured partitions: u-boot, u-boot-env, ubi

Signed-off-by: Konstantin Meshkov <skyline@mail.ru>
[ improve commit title ]
Link: https://github.com/openwrt/openwrt/pull/17924
Signed-off-by: Christian Marangi <ansuelsmth@gmail.com>

Neustradamus commented:
I have forgotten to thank @SimplyKot for this good PR and @Ansuel for merging!

Happy New Year to all!
Help is always welcome...

Have you progressed on DNS-320 and DNS-323 supports?
DNS-320: @APCCV, @wigyori
DNS-323: @glassum, @awanga, @BulldozerBSG

DNS-327L (badly no specs page on https://wikidevi.wi-cat.ru/ + https://techinfodepot.shoutwiki.com/)
But @nekromant has done the Debian support (I think, he can help for OpenWrt and to give all informations for wiki pages):
- https://ncrmnt.org/2015/02/15/debian-dlink-dns-327l-the-guide/
- https://ncrmnt.org/2014/10/28/dlink-dns-327l-how-you-should-not-make-a-nas/

DNS-343 (badly no specs page on https://wikidevi.wi-cat.ru/ + https://techinfodepot.shoutwiki.com/), it can be added with informations from here:
- http://dns323.kood.org/dns-343

D-Link DNS-320 rev A1
- http://dns323.kood.org/dns-320
- https://wikidevi.wi-cat.ru/D-Link_DNS-320_rev_A1
- https://techinfodepot.shoutwiki.com/wiki/D-Link_DNS-320_rev_A1

DNS-321 (badly no specs page on https://wikidevi.wi-cat.ru/ + https://techinfodepot.shoutwiki.com/), it can be added with informations from here:
- http://dns323.kood.org/dns-321

DNS-323
- http://dns323.kood.org/start
- https://wikidevi.wi-cat.ru/D-Link_DNS-323
- https://techinfodepot.shoutwiki.com/wiki/D-Link_DNS-323

-------------

D-Link NAS models:
- DNS-313
- DNS-315
- DNS-320
- DNS-320L
- DNS-320LW
- DNS-321
- DNS-323
- DNS-325
- DNS-327L
- DNS-340L
- DNS-343
- DNS-345

CHKDSK88 commented:
It was added after 24.10 branch:

https://firmware-selector.openwrt.org/?version=25.12.0-rc5&target=kirkwood%2Fgeneric&id=dlink_dns-325-a1

Taylor150 commented:
> It was added after 24.10 branch:
> 
> https://firmware-selector.openwrt.org/?version=25.12.0-rc5&target=kirkwood%2Fgeneric&id=dlink_dns-325-a1

Thank you!

Neustradamus commented:
@CHKDSK88: Thanks for your reactivity :)

I always search help to have other D-Link DNS-XXX NAS support...

SimplyKot commented:
I still can't solve the problem with the inverted LEDs on the network card. :-(
